kornia/geometry/liegroup/so3.py

```diff
@@ -0,0 +1,199 @@
+# kornia.geometry.quaternion module inspired by Sophus-sympy.
+# https://github.com/strasdat/Sophus/blob/master/sympy/sophus/so3.py
+from kornia.core import concatenate, stack, zeros_like, zeros
+from kornia.geometry.quaternion import Quaternion
+from kornia.testing import KORNIA_CHECK_TYPE, KORNIA_CHECK
+
+
+class So3:
+    r"""Base class to represent the So3 group.
+    The SO(3) is the group of all rotations about the origin of three-dimensional Euclidean space
+    R^3 under the operation of composition.
+    See more: https://en.wikipedia.org/wiki/3D_rotation_group
+
+    We internally represent the rotation by a unit quaternion.
+
+    Example:
+        >>> q = Quaternion.identity(batch_size=1)
+        >>> s = So3(q)
+        >>> s.q
+        real: tensor([[1.]], grad_fn=<SliceBackward0>) 
+        vec: tensor([[0., 0., 0.]], grad_fn=<SliceBackward0>)
+    """
+
+    def __init__(self, q: Quaternion) -> None:
+        """Constructor for the base class.
+
+        Args:
+            data: Quaternion with the sape of :math:`(B, 4)`.
+
+        Example:
+            >>> data = torch.rand((2, 4))
+            >>> q = Quaternion(data)
+            >>> So3(q)
+            real: tensor([[0.2734],
+                          [0.7782]], grad_fn=<SliceBackward0>) 
+            vec: tensor([[0.2420, 0.2716, 0.6159],
+                         [0.8727, 0.7592, 0.7212]], grad_fn=<SliceBackward0>)
+        """
+        KORNIA_CHECK_TYPE(q, Quaternion)
+        self._q = q
+        self.epsilon = 2.220446049250313e-16
+
+    def __repr__(self) -> str:
+        return f"{self.q}"
+
+    def __getitem__(self, idx):
+        return self._q[idx]
+
+    @property
+    def q(self):
+        """Return the underlying data with shape :math:`(B,4)`."""
+        return self._q
+
+    def exp(self, v) -> 'So3':
+        """Converts elements of lie algebra to elements of lie group.
+        See more: https://vision.in.tum.de/_media/members/demmeln/nurlanov2021so3log.pdf
+
+        Args:
+            v: vector of shape :math:`(B,3)`.
+
+        Example:
+            >>> v = torch.rand((2,3))
+            >>> s = So3.identity(batch_size=1).exp(v)
+            >>> s
+            real: tensor([[0.8891],
+                    [0.9774]], grad_fn=<SliceBackward0>) 
+            vec: tensor([[0.0893, 0.3285, 0.3060],
+                    [0.0567, 0.1315, 0.1558]], grad_fn=<SliceBackward0>)
+        """
+        theta = self.squared_norm(v).sqrt()
+        small_angles_mask = theta < self.epsilon
```

Comment:
Can you simplify this  exp and log a bit ? Use temporary variables to alleviate indexing and help with readability

Reply:
Not sure also about inplace assignments. Explore also using `where`